### PR TITLE
🌱 Tests cleanup

### PR DESCRIPTION
**Pull request by Adirio**
# Description

The following list defines all the checks and tests that we currently have, with some consideration on prerequisites and under which conditions are being executed (go version, devide OS/arch, and kubernetes version).

### Lint
- **Description:** verifies the code style.
- **Prerequisites:** none.
- **Make:** `make lint` to report the errors or `make lint-fix` to solve them.
- **CI:** yes (reporting, GH Actions).
- **Parameters:** single execution.

### Unit tests
- **Description:** unit tests defined in `*_test.go` files run by `go test ...` that test the different parts of the library isolated.
- **Prerequisites:** lint. *1
- **Make:** `make test-unit`.
- **CI:** yes (GH Actions & Prow). *2
- **Parameters:** multiple executions based on device OS/arch as right now. *3

### Coverage *4
- **Description:** obtains coverage info of unit tests and upload it to coveralls.
- **Prerequisites:** unit tests.
- **Make:** `make test-coverage` obtains the coverage info.
- **CI:** yes (GH Actions).
- **Parameters:** single execution.

### Integration tests *5
- **Description:** runs some basic kubebuilder commands to test the CLI.
- **Prerequisites:** unit tests.
- **File:** `test/integration.sh`.
- **Make:** `test-integration`.
- **CI:** yes (Prow).
- **Parameters:** single execution.

### Testdata check
- **Description:** verifies that the code under the `testdata` directory is up to date.
- **Prerequisites:** lint. *6
- **File:** `test/testdata/check.sh`.
- **Make:** `make check-testdata`.
- **CI:** yes (GH Actions).
- **Parameters:** single execution. *7

### Testdata tests
- **Description:** executes the unit tests of the scaffolded projects in the testdata directory.
- **Prerequisites:** integration tests. *8
- **File:** `test/testdata/test.sh`.
- **Make:** `make test-testdata`.
- **CI:** yes (Prow).
- **Parameters:** single execution. *7

### E2E tests
- **Description:** end-to-end tests.
- **Prerequisites:** none. *9
- **File:** `test/e2e/local.sh` or `test/e2e/ci.sh`.
- **Make:** `make test-e2e-local` or `make test-e2e-ci`.
- **CI:** yes (Prow).
- **Parameters:** multiple executions for several kubernetes versions.

## Notes:
1. Linting is not really a prerequisite for unit tests in the most strict sense, but as it is pretty fast, I don't think that it matters that much.
2. We should remove the overlap between GH Actions and Prow.
3. May be interesting to also test for multiple go versions if we do support more than one major go version.
4. We may be able to obtain the coverage info directly from one of the unit tests instead of reruning the tests again for coverage but this may be a bit tricky as we only need to upload one coverage report but unit tests are being executed for different OS/arch.
5. Integration tests are pretty old and have very few cases tested. Testdata tests below are also integrations tests that have way more cases tested and the generated code is actually checked. We should probably remove the old integration tests in favor of the testdata tests.
6. In order to check the testdata folder we need to regenerate it, which requires a running kubebuilder CLI, so testdata check should have as prerequisite the unit tests, not the lint.
7. It may be interesting to run testdata check and tests for several go versions and device OS/arch.
8. We should set check testdata as the prerequisite for test testdata, which would require to move this test to GH Actions.
9. I think that we should set as prerequisites all the previous tests (including unit tests, and testdata tests) which would require to move this tests to GH Actions.

### CODE REVIEW

File: test/testdata/generate.sh
Comment (camilamacedo86) [inline test/testdata/generate.sh:61]: if you remove it from here then it will no longer validate the --force in line 53.
So, it needs to be revert

File: test/testdata/generate.sh
Comment (camilamacedo86) [inline test/testdata/generate.sh:51]: if is make false then we cannot validate the --force properly since the manifest will not be gen.

File: test/testdata/generate.sh
Comment (camilamacedo86) [inline test/testdata/generate.sh:85]: btw, why you add the make=false in the above instructions? Was it done to solve so problem faced after the changes?

File: test/testdata/test.sh
Comment (camilamacedo86) [inline test/testdata/test.sh:45]: space ^nit

File: test/testdata/generate.sh
Comment (Adirio) [reply test/testdata/generate.sh:51]: The `--force` flag doesn't act different if there is or there isn't any manifest. It has to do with the tracked resources and the scaffolded files.

File: test/testdata/generate.sh
Comment (Adirio) [reply test/testdata/generate.sh:85]: There is no reason to compile the manager in the testdata folder and then remove it. There is no reason to go fmt and go vet all the directory for every creat call, etc. We already chose to set `--make=false` but some resources that were added after or webhooks which didnt have that flag previously still didnt have it, I just added it to the remaining ones.

File: test/testdata/test.sh
Comment (Adirio) [reply test/testdata/test.sh:45]: What space?

**Comment by Adirio** [comment]
I listed all the tests in the PR description and added several notes. As of now I didn't change the behavior of the tests, but I would like to know your opinion on the raised notes. I think 5 (removing out-dated integration tests) and those related to moving the tests to GH Actions and prerequisites (2, 6, 8 & 9) should be done. The rest (1, 3, 4 & 7) are just some comments that we may want to consider.

@estroz @DirectXMan12 @camilamacedo86

**Comment by Adirio** [comment]
As commented in the last kubebuilder/controller-runtime/controller-tools meeting, there are still some blockers that prevent us from using GH Actions without Prow. This means that notes 2, 8 and 9 need to wait until those blockers are solved. Point 5 and 6 could still be applied, and 1, 3, 4, & 7 remain being just some comments to bear in mind.

File: test/e2e/local.sh
Comment (estroz) [inline test/e2e/local.sh:2]: Nit: keep license/functional code separate
```suggestion

# Copyright 2018 The Kubernetes Authors.
```

File: test/testdata/test.sh
Comment (estroz) [inline test/testdata/test.sh:2]:
```suggestion

# Copyright 2018 The Kubernetes Authors.
```

File: pkg/plugins/golang/v2/scaffolds/internal/templates/api/types.go
Comment (estroz) [inline pkg/plugins/golang/v2/scaffolds/internal/templates/api/types.go:90]: Nit: make scaffolding changes in a separate PR

File: test/e2e/ci.sh
Comment (estroz) [inline test/e2e/ci.sh:2]:
```suggestion

# Copyright 2018 The Kubernetes Authors.
```

**Review by estroz — CHANGES_REQUESTED** [review]
Overall lgtm, just some nits

File: test/e2e/local.sh
Comment (Adirio) [reply test/e2e/local.sh:2]: Most of the files didnt have the space so I made all follow the same pattern but your comment makes sense, will add the space to all the files.

File: pkg/plugins/golang/v2/scaffolds/internal/templates/api/types.go
Comment (Adirio) [reply pkg/plugins/golang/v2/scaffolds/internal/templates/api/types.go:90]: Tests will break without this. These spaces were removed by a `go fmt` call that is no longer used.

**Comment by Adirio** [comment]
Rebased and applied one of the NITs and explained why the other needs to be in this PR and can't be done separately.

**Review by estroz — APPROVED** [review]
/approve

**Comment by Adirio** [comment]
@estroz `lgtm` left. Do you want another reviewer to take a look or did you just miss that label?

**Comment by estroz** [comment]
@Adirio yes since this is such a big PR I'd like another set of eyes for a lgtm.

**Comment by Adirio** [comment]
@DirectXMan12 I know Bash is terrible but can you take a look at this?

**Comment by camilamacedo86** [comment]
HI @Adirio, 

I think we just need to rebase this with the master branch to get it merged. Am I right? Please, feel free to do that and ping me.

**Comment by Adirio** [comment]
Yep, but if we add config/v3 we may need to modify the generate testdata file so I will rebase after that.

**Comment by Adirio** [comment]
Rebased. There is a single really small conflict with config/v3, so the order in which they are merged doesn't really matter.

File: testdata/project-v2-multigroup/go.mod
Comment (camilamacedo86) [inline testdata/project-v2-multigroup/go.mod:8]: Why a cleanup in the tests is change the go.mod?

Comment (Adirio) [reply testdata/project-v2-multigroup/go.mod:8]: This two requires are added to go mod when you run the unit tests of the projects. Previously `generate_tstdata.sh` was running the tests, now there is a specific target for running the tests and they are being run in the CI but not when generating the testdata. `test/testdata/generate.sh` just generates it, `test/testdata/check.sh` regenerates it and checks if there was any diff, and `test/testdata.test.sh` performs the tests in the testdata files. So as no tests are being run in generate these two packages are not `go get`ed and therefore not listed in the `go.mod` files.

Comment (camilamacedo86) [reply testdata/project-v2-multigroup/go.mod:8]: The generate_testdata.sh run make all for all generated project which is what prevent we introduce issues many times. For example, some scaffold issue that will affect a makefile target or missing the binaries to run the make tests target of the projects scaffolds. 

Could you please point out where the make all still been called for all testdata samples in order to allow the merge?

Comment (camilamacedo86) [reply testdata/project-v2-multigroup/go.mod:8]: I found: 

make test -> call integration-test which will run make https://github.com/kubernetes-sigs/kubebuilder/pull/1933/files#diff-21d992da4bf631674105f51a8e9ddf608cef06e0e53b568489c991e8f34ce32dR58

**Review by camilamacedo86 — COMMENTED** [review]
It shows fine for me 

/lgtm 

Great job. 🥇

File: testdata/project-v2-multigroup/go.mod
Comment (Adirio) [reply testdata/project-v2-multigroup/go.mod:8]: Exactly

**Pull request merged**
